libyul/backends/wasm/EVMToEWasmTranslator.cpp

```diff
@@ -0,0 +1,370 @@
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Translates Yul code from EVM dialect to eWasm dialect.
+ */
+
+#include <libyul/backends/wasm/EVMToEWasmTranslator.h>
+
+#include <libyul/backends/wasm/WordSizeTransform.h>
+#include <libyul/backends/wasm/WasmDialect.h>
+#include <libyul/optimiser/ExpressionSplitter.h>
+#include <libyul/optimiser/FunctionGrouper.h>
+#include <libyul/optimiser/MainFunction.h>
+#include <libyul/optimiser/FunctionHoister.h>
+#include <libyul/optimiser/Disambiguator.h>
+#include <libyul/optimiser/NameDisplacer.h>
+
+#include <libyul/AsmParser.h>
+#include <libyul/AsmAnalysis.h>
+#include <libyul/AsmAnalysisInfo.h>
+#include <libyul/Object.h>
+
+#include <liblangutil/ErrorReporter.h>
+#include <liblangutil/Scanner.h>
+#include <liblangutil/SourceReferenceFormatter.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+using namespace langutil;
+
+namespace
+{
+static string const polyfill{R"({
+function or_bool(a, b, c, d) -> r {
+	r := i64.ne(0, i64.or(i64.or(a, b), i64.or(c, d)))
+}
+function add_carry(x, y, c) -> r, r_c {
+	let t := i64.add(x, y)
+	r := i64.add(t, c)
+	r_c := i64.or(
+		i64.lt_u(t, x),
+		i64.lt_u(r, t)
+	)
+}
+function add(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
+	let carry
+	r4, carry := add_carry(x4, y4, 0)
+	r3, carry := add_carry(x3, y3, carry)
+	r2, carry := add_carry(x2, y2, carry)
+	r1, carry := add_carry(x1, y1, carry)
+}
+function bit_negate(x) -> y {
+	y := i64.xor(x, 0xffffffffffffffff)
+}
+function sub(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
+	// x - y = x + (~y + 1)
+	let carry
+	r4, carry := add_carry(x4, bit_negate(y4), 1)
+	r3, carry := add_carry(x3, bit_negate(y3), i64.add(carry, 1))
+	r2, carry := add_carry(x2, bit_negate(y2), i64.add(carry, 1))
+	r1, carry := add_carry(x1, bit_negate(y1), i64.add(carry, 1))
+}
+function split(x) -> hi, lo {
+	hi := i64.shr_u(x, 32)
+	lo := i64.and(x, 0xffffffff)
+}
+// Multiplies two 64 bit values resulting in a 128 bit
+// value split into two 64 bit values.
+function mul_64x64_128(x, y) -> hi, lo {
+	let xh, xl := split(x)
+	let yh, yl := split(y)
+
+	let t0 := i64.mul(xl, yl)
+	let t1 := i64.mul(xh, yl)
+	let t2 := i64.mul(xl, yh)
+	let t3 := i64.mul(xh, yh)
+
+	let t0h, t0l := split(t0)
+	let u1 := i64.add(t1, t0h)
+	let u1h, u1l := split(u1)
+	let u2 := i64.add(t2, u1l)
+
+	lo := i64.or(i64.shl(u2, 32), t0l)
+	hi := i64.add(t3, i64.add(i64.shr_u(u2, 32), u1h))
+}
+function add3_carry(a, b, c, carr) -> x, carry {
```

Comment:
I think `add3_carry(2^64-1, 2^64-1, 2^64-1, 0)` should return `2^64-3, 2` instead of `2^64-3, 1`? (considering how you use this function in `mul_128x128_256`)

Reply:
oh you are right, I should not pass the carry back in. Thanks!